feat(client): ✨ Add Win/Loss Screen

SBAsai opened the pull request:
Adds the win/loss screens for the players to view at the end of the game. Different screens are shown depending on the player's role and the game's outcome

SBAsai committed:
feat(client): ✨ Start Win/Loss Skeleton

SBAsai committed:
feat(client): ✨ Add cat puns

Add css to each win screen (and puns)

SBAsai committed:
fix(client): 🐛 Fix win loss functions

Updated the variables to fit that of the documents in firebase

SBAsai committed:
Merge branch 'development' into win-loss

SBAsai committed:
fix(client) :bug: Update if statements

Updated the if statements to use the proper variables and grab from the player documents

SBAsai committed:
fix(client): 🐛 Fix array errors and player data retreival

Fixed errors with the alive status of each player showing up as invalid

SBAsai committed:
Merge branch 'development' into win-loss

SBAsai committed:
fix(client): 🐛 Fix retreiving user's role

Fixed an issue with retreiving the player's role from the private player collection

nstringham reviewed (changes requested):
can you pull from development please

SBAsai committed:
Merge branch 'development' into asai/win-loss

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+        }
+      }
+    }
+  }
```
missing `)`

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+  let lobbyRoles: PrivatePlayer; // cycle through the lobby and retrieve their roles
+
+  function lobbyReturn() {
+      // TODO: change gamestate back to lobby
```
we probobly don't want to do that on the client side

SBAsai replied:
Does this mean I should make the function under the functions folder or under lib/firebase?

nstringham replied:
it should be in firebase functions aka `/functions/src/` aka server side

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+  }
+  onMount (async () => {
+    const lobby = doc(lobbyCollection, code);
+    const validLobby = await getDoc(lobby);
```
why not use `export let`?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+  let aliveCatCount = 0;
+  let catfishDisplayname = "";
+  let currentRole: PrivatePlayer; // the player's current role
+  let lobbyRoles: PrivatePlayer; // cycle through the lobby and retrieve their roles
```
why is this not scoped to the loop where it is used?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+          aliveCatCount++;
+          // within the alive players, count the number that are cats
+        } else {
+          catfishDisplayname = players[i].displayName;
```
what if there is more than one catfish

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+    let aliveCatFishCount = aliveCount - aliveCatCount;
+    // any other player who is alive and not a cat is a catfish
+    if ($user !== null) {
+      const currentPlayer = await getDoc(doc(getPrivatePlayerCollection(lobby), $user.uid));
+      currentRole = currentPlayer.data() as PrivatePlayer;
```
why not use `export let`

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+
+  export let code: string;
+
+  let end: string = "";
```
the type of this should probobly be enumerated
```ts
let end: "Cat Win" | "Cat Lose" | "Catfish Lose" | "Catfish Win" | null = null;
```
or something like that

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,168 @@
+      <h2>{catfishDisplayname} was sentenced to 9 lives in purrison!</h2>
+      <button on:click={lobbyReturn}>Return to Lobby</button>
+    </div>
+  {/if}
+  {#if (end = "Cat Lose")}
```
it would probobly be better to use else if

nstringham reviewed (changes requested):
this component needs to be imported and used in the game route

SBAsai committed:
style(client): 🎨 Fix variable scopes and implement export lets

Fixed the variables not being scoped inside the loops. Also used export lets for various data grabbing from documents. "Return to Lobby" button now functions to change gamestate to "WAIT". Updated game page to include the "END" game state.

SBAsai committed:
style(client): 🎨 Fix prettier warning

Formatted winloss.ts with prettier

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,164 @@
+
+    <div class="displayname">
+      <h2>You should go back to your sea ani-anim-aneme... home</h2>
+      <button on:click={()=>lobbyReturn}>Return to Lobby</button>
```
this does not invoke lobbyReturn

nstringham replied:
this does not invoke lobbyReturn

nstringham commented on client/src/lib/firebase/winloss.ts:
```diff
@@ -0,0 +1,29 @@
+  return currentPlayer.data();
+};
+
+export const lobbyReturn = async (code: string) => {
+  const lobbyData = await getLobbyData(code);
+  if (lobbyData !== undefined) {
+    lobbyData.state = "WAIT";
+  }
+};
```
this doesn't actually do anyhting because the variable `lobbyData` is garbage collected after going out of scope so the state change is lost forever.

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,164 @@
+              aliveCatCount++;
+              // within the alive players, count the number that are cats
+            } else {
+              catfishDisplayname[i] = lobbyData.players[i].displayName; // create a list of catfishes
```
doesn't this create a array with holes in it?

nstringham replied:
doesn't this create a array with holes in it?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,164 @@
+      }
+    }
+    let aliveCatFishCount = aliveCount - aliveCatCount;
+    catfishList = [catfishDisplayname.slice(0, -1).join(', '), catfishDisplayname.slice(-1)[0]].join(catfishDisplayname.length < 2 ? '' : ' and ');
```
you might want to consider `Intl.ListFormat()` which makes english lists for you https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/ListFormat

nstringham commented on client/src/lib/firebase/winloss.ts:
```diff
@@ -0,0 +1,29 @@
+  return currentPlayer.data();
+};
+
+export const lobbyReturn = async (code: string) => {
+  const lobbyData = await getLobbyData(code);
+  if (lobbyData !== undefined) {
+    lobbyData.state = "WAIT";
+  }
+};
```
this doesn't actually do anyhting because the variable `lobbyData` is garbage collected after going out of scope so the state change is lost forever.

SBAsai replied:
Would I want to use setDoc here instead to change the game state?

nstringham replied:
well yes `setDoc ` would work but it would probably be better to do this on the server in firebase functions

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,164 @@
+    catfishList = [catfishDisplayname.slice(0, -1).join(', '), catfishDisplayname.slice(-1)[0]].join(catfishDisplayname.length < 2 ? '' : ' and ');
+    // any other player who is alive and not a cat is a catfish
+    if ($user !== null) {
+        currentRole = await getCurrentPlayer(lobbyCode, $user.uid); // grab the current user's role
```
I really think `export let` would be better here because we already have that data in the game page

SBAsai committed:
style(client): 🎨 Use export lets, update game page, update catfish array

Implemented the use of export lets for various variables. Updated the catfish array with a separate counter to avoid holes. Changed formatting on the catfishList string.

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,163 @@
+      for (let i = 0; i < lobby.uids.length; i++) {
+        if (lobby.players[i].alive == true) {
+          // count the number of remaining players who are alive
+          lobbyRoles = await getPrivatePlayer(lobbyCode, lobby.uids[i]);
```
you can't read someone elses private player so this code should probobly be on the server side

nstringham commented on client/src/lib/firebase/winloss.ts:
```diff
@@ -0,0 +1,30 @@
+export const getCurrentPlayer = async (code: string, id: string) => {
+  const currentPlayer = await getDoc(doc(getPrivatePlayerCollection(getLobby(code)), id));
+  return currentPlayer.data();
+};
```
this function is identical to getPrivatePlayer

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,163 @@
+              aliveCatCount++;
+              // within the alive players, count the number that are cats
+            } else {
+              catfishDisplayname[catfishCounter] = lobby.players[i].displayName; // create a list of catfishes
+              catfishCounter++;
```
did you know about `push`?
what you did works but push will make your life easier if you didn't already know about it.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push

SBAsai committed:
style(client): 🎨 Move functions to server side

-Moved the functions from lib/firebase to the server side
- Fixed invoking the lobbyReturn method
- Implemented array.push()
-Fixed game state change function

SBAsai committed:
fix(client): 🐛 Fix import statement

SBAsai commented:
Lint functions is failing because of import statements, but I cannot figure out why it's throwing an error

SBAsai committed:
style(client): 🎨 Fix server-sided functions

-Attempted to implement a lobbyReturn function with onCall() and fetch private player data using transaction. Also moved role grabbing for-loop to the server.

SBAsai committed:
style(client): 🎨 Fix prettier, fix for-loop variable declaration

Formatted with prettier and created return statement on for-loop to retrieve the variables

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
fix(client): 🐛 Fix import statement

Fixed firebase-functions import call

SBAsai committed:
feat(client): ✨ Implement functioning win/loss page

Fixed serveri-sided functions to properly read from the private player doc. Fixed client-sided code to properly take data from the server-sided functions and display it onto the html. Implemented error response for grabbing the lobby's roles.

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,184 @@
+      if (catfishCount == 0) {
+        return "Cat Win";
+      }
+      if (catCount == catfishCount) {
+        return "Cat Lose";
+      }
+      if (catCount < catfishCount) {
+        return "Cat Lose";
+      }
```
this should proboly be an else case

nstringham replied:
this should proboly be an else case

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,184 @@
+        return "Cat Lose";
+      }
+    } else {
+      if (catCount == 0) {
+        return "Catfish Win";
+      }
+      if (catfishCount == catCount) {
+        return "Catfish Win";
+      }
+      if (catfishCount == 0) {
+        return "Catfish Lose";
+      }
```
what if there are 2 catfish and 1 cat?

nstringham replied:
what if there are 2 catfish and 1 cat?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,184 @@
+
+  export let lobbyCode: string;
+  export let privatePlayer: PrivatePlayer | undefined;
+  export let lobby: Lobby | undefined;
```
I don't think these will ever be undefined

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,68 @@
+  return privatePlayer.data();
+}
+
+export const getLobbyRole = functions.https.onCall(async (data: unknown, context) => {
```
this should probobly not be run for every client. it should run only once and then the results should be put in the database so everyone can see it

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,184 @@
+      {/if}
+      {#if lobby !== undefined}
+        <button
+          on:click={async () => {
```
you don't need async here

SBAsai replied:
I'm doing it the same way start game is doing it so wouldn't the button invoking be the same here too?

nstringham replied:
it's also unnecessary on the start game button thanks for pointing that out

SBAsai committed:
style(client): 🎨 Fix structure of if statements and variable calls.

Fixed missing scenarios for cat and catfish within the if statements. Removed "undefined" from a couple variables. Added a check for counting the non-alive catfish since it was not checking that before.

SBAsai committed:
style(client): 🎨 Fix prettier warning

Formatted to prettier

SBAsai committed:
style(client): 🎨 Fix extra async function call

Removed async from component function call

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
feat(client): ✨ Change server-side code implementation and add post-lobby function

Changed the win/loss screen role calculation to only be run once in the server and store data into the database. Also made sure players are now set to alive once they are returned to lobby

SBAsai committed:
Merge branch 'development' into asai/win-loss

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,26 @@
+  if (context.auth.uid !== uids[0]) {
+    return { error: "Not allowed to return to lobby!" };
+  }
+  return lobbyCollection.doc(data.code).update({ state: "WAIT" });
```
we should also some code to delete private players and set everyone to alive and maybe some other stuff I am forgeting

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,26 @@
+  if (lobby.exists === false) {
+    return { error: "Lobby doesn't exist!" };
+  }
+
```
we should probobly also check to see if the state of the game is `END`

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -24,6 +24,11 @@ export type Player = {
+  /**
+   * the role that the player played; used at the end of the game
+   */
+  role: Role;
```
this should probobly be optional

SBAsai replied:
Sorry can you expand on this? Is there an optional type in Typescript I'm blanking on?

nstringham commented on functions/src/lobby.ts:
```diff
@@ -185,3 +193,74 @@ export const collectPromptAnswers = functions.firestore
+  }
+  let catCount = 0;
+  let catfishCount = 0;
+  const catfishDisplayname: string[] = [];
```
looks like `catfishDisplayname` is never used

nstringham commented on functions/src/lobby.ts:
```diff
@@ -185,3 +193,74 @@ export const collectPromptAnswers = functions.firestore
+    }
+  }
+
+  let winner: Role | undefined;
```
this should probobly just be `Role`

nstringham commented on functions/src/lobby.ts:
```diff
@@ -97,6 +97,14 @@ export const onLobbyUpdate = functions.firestore.document("/lobbies/{code}").onU
     await startPrompt(lobbyDocRef);
   }
+
+  if (lobby.state == "END" && oldLobby.state != "END") {
+    addRole(lobbyDocRef);
+  }
```
what sets the state to `END`?

SBAsai replied:
I'm assuming voting will have a function that will change the lobby state  right?

nstringham replied:
Hmm you and @cmcclusky should talk and make sure you aren't each assuming that eachother is going to do that

SBAsai replied:
For sure, I'll message him

nstringham commented on functions/src/lobby.ts:
```diff
@@ -185,3 +193,74 @@ export const collectPromptAnswers = functions.firestore
+  }
+
+  let winner: Role | undefined;
+  if (catfishCount == 0) {
+    winner = "CAT";
+  }
+  if (catCount <= catfishCount) {
+    winner = "CATFISH";
+  }
```
these do not cover every case

SBAsai replied:
I'm honestly struggling to figure out what case I'm missing. `if(catCount == 0)`?

nstringham replied:
For example if there is 1 catfish and 3 cats.

I feel like this will be more obvious once you integrate with the voting code

SBAsai replied:
If there’s 1 catfish and 3 cats the game would continue right? I think neither role wins yet in that situation

nstringham commented on client/src/lib/firebase/firestore-functions.ts:
```diff
@@ -1,8 +1,10 @@
 export const joinLobby = httpsCallable<LobbyRequest, ErrorResponse>(functions, "joinLobby");
+export const lobbyReturn = httpsCallable<LobbyRequest, ErrorResponse>(functions, "lobbyReturn");
 export const changeAvatar = httpsCallable<ChangeAvatarData, void>(functions, "changeAvatar");
+export const getLobbyRole = httpsCallable<LobbyRequest, LobbyRoleResponse | ErrorResponse>(functions, "getLobbyRole");
```
looks like this doesn't exist

nstringham commented on functions/src/firestore-functions-types.ts:
```diff
@@ -6,3 +6,8 @@ export function isLobbyRequest(data: unknown): data is LobbyRequest {
   return data != null && typeof data === "object" && "code" in data && typeof (data as LobbyRequest).code === "string";
 }
+
+export type LobbyRoleResponse = {
+  aliveCatCount: number;
+  catfishDisplayname: string[];
+};
```
looks like this is never used

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -28,7 +31,7 @@
     }
 
-    const lobbyDocRef = doc(lobbyCollection, lobbyCode);
+    lobbyDocRef = doc(lobbyCollection, lobbyCode);
```
why is this changed from const to let?

SBAsai replied:
I used lobbyDocRef in an export let for my own WinLoss.svelte component so I had to change that

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,201 @@
+    catfishList = formatter.format(catfishes);
+
+    // grab the current user's role
+    const myRole = (await getDoc(doc(getPrivatePlayerCollection(lobbyDocRef), $user?.uid))).data() as PrivatePlayer;
```
this is now on the lobby so we don't need to do an extra database read

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,201 @@
+        <h2>{catfishList} was sentenced to 9 lives in purrison!</h2>
+      {/if}
+      <!-- {#if isHost} -->
+      <button
+        on:click={() => {
+          lobbyReturn({ code: lobbyCode });
+        }}>Return to Lobby</button
+      >
```
in `winloss.ts` you say:
```ts
if (context.auth.uid !== uids[0]) {
  return { error: "Not allowed to return to lobby!" };
}
```
if that's how you are going to make it work then this button should only be visable to the host

SBAsai replied:
I forgot to comment it out, but since we're adding timers now, the lobby return button is not going to be used. I just have the lobbyReturn function to test until the timers get added

nstringham replied:
Gotcha.

SBAsai committed:
feat(client): ✨ Add deletion of the private player collection upon game end

Added a function that deletes the private player collection after the players return to the lobby. Fixed various small styling errors within the code.

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
style(client): 🎨 Add optional question mark

Changed the location of the optional question mark

SBAsai committed:
style(client): 🎨 Made winner optional

Added optional question mark to winner in firestore types

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+    winner = "CATFISH";
+  }
+
+  return lobbyDocRef.update({ players, winner });
```
this should use a transaction because we are reading data and then writing it back

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+export async function addRole(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    return { error: "Lobby doesn't exist!" };
```
this should be a throw

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+export async function makeAlive(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    return { error: "Lobby doesn't exist!" };
```
this should be a throw

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+  // loop through each player and set their alive status to true
+  const { players } = lobby.data() as Lobby;
+  for (let i = 0; i < players.length; i++) {
+    players[i].alive = true;
```
should probobly also remove role from player

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+  for (let i = 0; i < players.length; i++) {
+    players[i].alive = true;
+  }
+  return lobbyDocRef.update({ players });
```
this should also be a transaction

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+
+  // loop through each doc in the private player collection via uid and delete them one-by-one
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= docCount; i++) {
```
why not use `uids.length` instead of `docCount`

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    return { error: "Lobby does not exist!" };
```
this should probobly be a throw

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+    deleted++;
+  }
+
+  if (deleted >= docCount) {
+    return;
+  }
```
what does this do?

SBAsai replied:
I honestly just took stuff from the documentation that looked important, but I think the original use case was having a check with a recursive call after the return. I don't think it's necessary here though so I'll remove it

nstringham commented on functions/src/lobby.ts:
```diff
@@ -213,6 +223,89 @@ export const collectPromptAnswers = functions.firestore
+  return lobbyDocRef.update({ players });
+}
+
+export async function deletePrivatePlayers(
```
it might be a good idea to querry instead of using uids from lobby but I'm not sure

SBAsai replied:
If it becomes an issue I'll look further into it

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,27 @@
+import { Lobby } from "./firestore-types/lobby";
+
+// this function will likely be obsolete once timers are introduced
+export const lobbyReturn = functions.https.onCall(async (data: unknown, context) => {
```
lets put an explict return type here like 191b0bf80c59858ed7b9e1b52d2e04ce9b9f2f9f

SBAsai replied:
I commented in line 6 that this function probably won't be used in the future, do you want me to just remove it completely now or when the timers are actually implemented and fix it now?

nstringham replied:
that depends on how close timers are to being done

SBAsai replied:
For now I'll just keep the function updated with the new changes

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,27 @@
+// this function will likely be obsolete once timers are introduced
+export const lobbyReturn = functions.https.onCall(async (data: unknown, context) => {
+  if (context.auth === undefined) {
+    return { error: "Not Signed In" };
```
please use errors like in aea453b216d98c6eef8187ee8791a3f67e9acc9f

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,27 @@
+    return { error: "Not allowed to return to lobby!" };
+  }
+  return lobbyCollection.doc(data.code).update({ state: "WAIT" });
+});
```
I feel like the lobby reseting stuff should be done here instead of in `onLobbyUpdate`

SBAsai replied:
Would I create a function similar to `onLobbyUpdate` or is there a separate way to call the functions in my own .ts file?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,197 @@
+    </div>
+
+    <div class="image">
+      <img src="https://www.azpetvet.com/wp-content/uploads/Happy-cat-1024x711.jpg" alt="clown fish" />
```
do we have rights to this image?

SBAsai replied:
I'm just kind of using these images as placeholders for this sprint. I'm not entirely sure if we've come to a decision on what we want here yet

SBAsai replied:
I'm just kind of using these images as placeholders for this sprint. I'm not entirely sure if we've come to a decision on what we want here yet

SBAsai replied:
I went ahead and just drew some pictures in paint so let me know how you feel about them

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,197 @@
+    </div>
+
+    <div class="image">
+      <img src="https://pbs.twimg.com/media/EAmr-PAWsAEoiWR.jpg" alt="cat fish" />
```
do we have rights to this image?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,197 @@
+    </div>
+
+    <div class="image">
+      <img src="https://media.fisheries.noaa.gov/2020-10/blue%20catfish%20face.jpg?VersionId=null" alt="cat fish" />
```
do we have rights to this image?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,197 @@
+    </div>
+
+    <div class="image">
+      <img src="https://cdn.drawception.com/drawings/577991/0C653P5Wka.png" alt="clown fish" />
```
do we have rights to this image?

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,197 @@
+  {:else if end == "Cat Lose"}
+    <div class="banner">
+      <h2>You have cat to be kitten me!</h2>
+      <h2>The impawster was not caught!</h2>
```
it would be nice to say who they were

nstringham reviewed (changes requested):
css needs some work but I can fix that later

SBAsai committed:
feat(client): ✨ Add new images, move function location

Added new (wonderfully hand-crafted) images for the win/loss pages and moved the win/loss functions from lobby.ts to winloss.ts. Re-introduced the lobby return button.

SBAsai committed:
fix(client): 🐛 Fix lint error

Fix import statement causing lint error

SBAsai committed:
Merge branch 'development' into asai/win-loss

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+    if (lobby === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
```
this isn't required

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+      throw new Error("Lobby does not exist!");
+    }
+    const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+    const { players, uids } = lobby.data() as Lobby;
```
this should be `lobbyData` instead of calling `lobby.data()` agian

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+        }
+      }
+    }
+    console.log(lobbyData.winner);
```
let's not use console.log in functions on production

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+    } else {
+      lobbyData.winner = "CATFISH";
+    }
+    console.log(lobbyData.winner);
```
let's not use console.log in functions on production

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+    });
+
+    // delete the private player collection
+    privatePlayerSnapshot.docs.map((privatePlayerDoc) => {
+      privatePlayerDoc.ref.delete();
+    });
```
this should probobly be a for loop

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,130 @@
+  });
+}
+
+export async function deletePrivatePlayers(
+  lobbyDocRef: firestore.DocumentReference<Lobby>,
+  privatePlayerCollectionRef: firestore.CollectionReference<PrivatePlayer>
+) {
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    throw new Error("Lobby does not exist!");
+  }
+
+  // loop through each doc in the private player collection via uid and delete them one-by-one
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= uids.length; i++) {
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    await playerDoc.delete();
+  }
+  return;
+}
```
this is unused

SBAsai committed:
style(client): 🎨 Remove unnecessary code, add for-loop

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
fix(client): 🐛 Fix import errors

Fixed the import locations

gchamb commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,217 @@
+    // grab the current user's role
+    const myRole = privatePlayer.role;
+    // update end variable depending on the winner type and the current player's role
+    if (myRole == "CAT") {
+      if (lobby.winner == "CAT") {
+        end = "Cat Win";
+      } else if (lobby.winner == "CATFISH") {
+        end = "Cat Lose";
+      }
+    } else {
+      if (lobby.winner == "CAT") {
+        end = "Catfish Lose";
+      } else if (lobby.winner == "CATFISH") {
+        end = "Catfish Win";
+      }
+    }
```
should check if `lobby.winner` is `undefined` before so you can just have if/else instead of elseif

gchamb commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,217 @@
+  {:else if end == "Cat Lose"}
+    <div class="banner">
+      <h2>You have cat to be kitten me!</h2>
+      {#if catfishes.length > 1}
+        <h2>The impawster, {catfishList} was not caught!</h2>
+      {:else}
+        <h2>The impawsters, {catfishList} were not caught!</h2>
+      {/if}
```
if catfishes.length is equal to 1 it will display `The impawsters, {catfishList} were not caught` because 1 isnt greater than 1

gchamb commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,217 @@
+        <h2>{catfishList} was sentenced to 9 lives in purrison!</h2>
+      {/if}
+      <!-- If the player is the host, show the return button -->
+      {#if isHost}
+        <button
+          on:click={() => {
+            lobbyReturn({ code: lobbyCode });
+          }}>Return to Lobby</button
```
you write this multiple times. Can you not render this once?

SBAsai replied:
How would I render it once? An event listener?

nstringham replied:
you can use `on:click|once` like in #107

SBAsai replied:
Oh I think Giantte wanted the whole if statement to only be written once instead of clicked once. It should be fixed.

nstringham replied:
Oh ok

SBAsai committed:
style(client): 🎨 Fix various minor issues

Fix multiple button renders, updated if statement to check winner, fix grammar issue

SBAsai committed:
fix(client): 🐛 Fix component not rendering properly

Fixed the component not rendering properly by changing onMount to $:

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,122 @@
+
+  // loop through each doc in the private player collection via uid and delete them one-by-one
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= uids.length; i++) {
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    await playerDoc.delete();
+  }
```
an `await` insode a loop can be really slow maybe consider a batch write https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,122 @@
+      throw new Error("Lobby does not exist!");
+    }
+    const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+    const { players, uids } = lobbyData as Lobby;
```
don't need this `as`

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,122 @@
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= uids.length; i++) {
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    await playerDoc.delete();
```
an `await` inside a loop can be really slow maybe consider a batch write https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,122 @@
+
+    for (let i = 0; i < players.length; i++) {
+      const privatePlayerDocRef = privatePlayerCollection.doc(uids[i]);
+      const privatePlayerDoc = await transaction.get(privatePlayerDocRef);
```
an `await` inside a loop can be really slow maybe consider using [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) and [Array.map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,122 @@
+      const privatePlayerData = privatePlayerDoc.data();
+
+      if (!privatePlayerData) {
+        throw new Error(`Private player not found for uid ${uids[i]}`);
```
where are we catching this error?

SBAsai replied:
um I'm throwing the same error as the one in `lobby.ts` so presumably it doesn't need a catch perhaps?

nstringham replied:
it probably shouldn't be like that in lobby either but I've bean thinking about this more and I think this is ok for now. we can fix it in sprint 3

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,204 @@
+  let isHost: boolean;
+  let myRole: Role;
+
+  $: {
```
I don't think you want one bug `$:` try to make them as small as posible so they only contain one reactive thing.

for example you could have one that sets the value of `isHost`

SBAsai committed:
style(functions): 🎨 Use batch write instead of transaction within for loop, use array map instead of other for loop

Implemented a write batch within the for loop instead of using a transaction every time. Used a Promise.all() and Array.map() instead of a for loop to set the public player role. Fixed some minor css issues.

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,125 @@
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    // delete each private player document with the WriteBatch
+    batch.delete(playerDoc);
+    batch.commit();
```
the commit should be outside the loop and should be awaited

nstringham commented on functions/src/lobby.ts:
```diff
@@ -149,12 +150,17 @@ export const onLobbyUpdate = functions.firestore.document("/lobbies/{code}").onU
   }
+
+  if (lobby.state == "END" && oldLobby.state != "END") {
+    findWinner(lobbyDocRef);
```
this needs to be awaited

nstringham replied:
this needs to be awaited

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,125 @@
+
+  // check game state
+  if (state !== "END") {
+    throw new functions.https.HttpsError("invalid-argument", "Wrong game state!");
```
`invalid-argument` is fine but in the future `failed-precondition` would probobly be better

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,125 @@
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    // delete each private player document with the WriteBatch
+    batch.delete(playerDoc);
+    batch.commit();
```
the commit should be outside the loop and should be awaited

SBAsai replied:
I originally had the commit outside the loop, but the delete wasn't being committed

nstringham replied:
> I originally had the commit outside the loop, but the delete wasn't being committed

did you forget to await the commit?

SBAsai replied:
no, I'm testing it right now too and it's not committing the deletes

SBAsai replied:
nvm I had `i <= uids.length` in my for loop and had to remove the `=` lmao

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,125 @@
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobby === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
```
this will never happen because transaction.get never returns undefined

nstringham commented on functions/src/winloss.ts:
```diff
@@ -0,0 +1,125 @@
+    throw new functions.https.HttpsError("invalid-argument", "Wrong game state!");
+  }
+
+  await resetLobby(lobby.ref);
+  await deletePrivatePlayers(lobby.ref, getPrivatePlayerCollection(lobby.ref));
```
each of these functions get the lobby doc wich causes a total of 3 reads. instead lobby should be passed into these functions

SBAsai replied:
`getPrivatePlayerCollection` requires a lobby ref document, so would it be fine to leave that as is?

nstringham replied:
> `getPrivatePlayerCollection` requires a lobby ref document, so would it be fine to leave that as is?

I'm not 100% sure what you're asking.

I think you're asking if it is ok to pass lobby and keep `getPrivatePlayerCollection(lobby.ref)` the way it is

and if that's what you are asking then yes

SBAsai replied:
yup, that was all I was wondering. Should be good then.

nstringham commented on client/src/components/WinLoss.svelte:
```diff
@@ -0,0 +1,202 @@
+  catfishList = formatter.format(catfishes);
+
+  // grab the current user's role
+  myRole = privatePlayer.role;
```
this is not reactive but honestly it would probobly be better to get rid of myRole entirly and just use privatePlayer.role on line 39

SBAsai committed:
fix(functions): 🐛 Fix batch commit inside for loop, remove multiple document reads

Moved batch.commit to be outside the for loop. Changed document read calls to not occur as often. Removed unnecessary code.

SBAsai committed:
Merge branch 'development' into asai/win-loss

SBAsai committed:
fix(client): 🐛 Add missing reactive call

Added a reactive call to catfishlist since it was not displaying the names of the catfish when the loss screen appears

gchamb reviewed (approved):
looks good to me. you addressed nate's changes
